src/derparams.f90

```diff
@@ -1,10 +1,667 @@
 module m_derparams
-   use m_common, only: dp
+   use m_common, only: dp, pi
 
    implicit none
 
+   type :: tdsops_t
+      real(dp), allocatable, dimension(:) :: dist_fw, dist_bw, dist_sa, dist_sc
+      real(dp), allocatable :: coeffs(:), coeffs_t(:, :), coeffs_b(:, :)
+      integer :: n, n_halo
+      real(dp) :: alpha, delta
+   contains
+      procedure :: deriv_1st, deriv_2nd, interpl_mid, stag_der_1st, preprocess
+   end type tdsops_t
+
+   interface tdsops_t
+      module procedure tdsops
+   end interface tdsops_t
+
 contains
 
+   type(tdsops_t) function tdsops(n, delta, operation, variation, n_halo, &
+                                  from_to, bc_top, bc_bot, paire, c_nu, nu0_nu)
+      implicit none
+
+      integer, intent(in) :: n
+      integer, optional, intent(in) :: n_halo
+      real(dp), intent(in) :: delta
+      character(*), intent(in) :: operation, variation
+      character(*), optional, intent(in) :: from_to, bc_top, bc_bot
+      logical, optional, intent(in) :: paire
+      real(dp), optional, intent(in) :: c_nu, nu0_nu
+
+      integer :: n_stencil
+
+      tdsops%n = n
+      tdsops%delta = delta
+      if (present(n_halo)) then
+         tdsops%n_halo = n_halo
+      else
+         tdsops%n_halo = 4
+      end if
+
+      n_stencil = 2*tdsops%n_halo + 1
+
+      allocate(tdsops%dist_fw(n), tdsops%dist_bw(n))
+      allocate(tdsops%dist_sa(n), tdsops%dist_sc(n))
+      allocate(tdsops%coeffs(n_stencil))
+      allocate(tdsops%coeffs_t(tdsops%n_halo, n_stencil))
+      allocate(tdsops%coeffs_b(tdsops%n_halo, n_stencil))
+
+      if (operation == 'first-deriv') then
+         call tdsops%deriv_1st(variation, bc_top, bc_bot, paire)
+      else if (operation == 'second-deriv') then
+         call tdsops%deriv_2nd(variation, bc_top, bc_bot, paire, c_nu, nu0_nu)
+      else if (operation == 'interpolate') then
+         call tdsops%interpl_mid(variation, from_to, bc_top, bc_bot, paire)
+      else if (operation == 'staggared-deriv') then
+         call tdsops%stag_der_1st(variation, from_to, bc_top, bc_bot, paire)
+      else
+         error stop 'operation is not defined'
+      end if
+
+      !if ((bc_top == 'periodic' .and. bc_bot /= 'periodic') .or. &
+      !    (bc_top /= 'periodic' .and. bc_bot == 'periodic')) then
+      !   error stop 'In case of periodic BCs both ends must be periodic'
+      !end if
+
+   end function tdsops
+
+   subroutine deriv_1st(self, variation, bc_top, bc_bot, paire)
+      implicit none
+
+      class(tdsops_t), intent(inout) :: self
+      character(*), intent(in) :: variation
+      character(*), optional, intent(in) :: bc_top, bc_bot
+      logical, optional, intent(in) :: paire
```

Comment:
Worth commenting procedure arguments

- what is variation? what values could it take?
- what are bc_top/bot? Are they related to the y-orientation or the top/bottom of the vector when viewed as a system of equations?
- what is `paire` (I know from Xcompact3d familiarity, but i think we should make this clear for the future)?

Reply:
- variation is renamed as scheme
- bc_top/bot was indicating y direction indeed, so renamed as bc_start and bc_end, other places in the framework may refer to this in short form such as .._b/.._e.
- paire is renamed as sym.
And describtion for all this is provided in the class constructor function.